### ISSUE TITLE
OverlappingMarkerSpiderfier is not defined when Google Maps Api are loaded in Asynchronous

### ISSUE THREAD

**Issue by Aminta** [jawj/OverlappingMarkerSpiderfier#90]
Hi!

When I load Gmaps API via "official" Asynchronous Loading (https://developers.google.com/maps/documentation/javascript/examples/map-simple-async) or via other plugins like Lazy Loading Gmaps (http://osvaldas.info/lazy-loading-google-maps) I got "OverlappingMarkerSpiderfier is not defined" or "Cannot read property 'HYBRID' of undefined" errors even if the oms file is actually loaded after the Gmaps Api.

**Comment by Jaans** [comment]
We had a similar issue. The Google Maps API must finish loading before OMS can load.

We use the Google jsAPI (http://www.google.com/jsapi) to load Google Maps API asynchronously, so what we did is load OMS in the onload callback.
```
if (window.addEventListener)
    window.addEventListener("load", loadOMS, false);
else if (window.attachEvent)
    window.attachEvent("onload", loadOMS);
else
    window.onload = loadOMS;
```

You could try something similar.

**Comment by Aminta** [comment]
Thanks, Jaans!

**Comment by IvikGH** [comment]
Hi, Jaans.
As I've understood `loadOMS` is a function.
Could you show it please ?

**Comment by Jaans** [comment]
@IvikGH Yes, it's just a basic function. This function gets called upon the load event (with some fallback implementation) - this event fortunately happens after the Google jsAPI finishes loading.

With the implementation of the function, we've chosen to inline the OMS script and load it into the dom. See below.
```
<script>
            function loadOMS() {
            	var element = document.createElement("script");
            	element.text = "<<OMS SCRIPT HERE>>";
            	document.body.appendChild(element);
            }
            
            if (window.addEventListener)
            	window.addEventListener("load", loadOMS, false);
            else if (window.attachEvent)
            	window.attachEvent("onload", loadOMS);
            else
            	window.onload = loadOMS;
</script>
```

**Comment by IvikGH** [comment]
Thanks.

**Comment by Frank72** [comment]
Hi Jaans,
I am loading my map async and the above solution does not work for me, am still getting **Uncaught ReferenceError: OverlappingMarkerSpiderfier is not defined**. I have also tried to add the debugging line to your code, after **var h=!0,u=null,v=!1;**

if (typeof google === 'object' && typeof google.maps === 'object') {console.log('1.loaded!');}else{console.log('1.not loaded!');}

it prints '1. loaded!' but the error remains the same, on the line 

oms = new OverlappingMarkerSpiderfier(map,{markersWontMove: true, markersWontHide: true});

could it be still because of async pls?

Tx

**Comment by SatdayaStudios** [comment]
I'm running into this async problem too, on loading google maps asynchronously. The above mentioned fix didn't work for me. Anyone know how to ensure a script like OMS loads before Google asynchronously loads?

**Comment by willbuckingham** [comment]
I was having similar issues, what worked for me was removing the callback=initMap from the google maps api loading, and removed the **async defer**
`<script src="https://maps.googleapis.com/maps/api/js?key=<?=GOOGLE_API_KEY;?>"></script><!--&callback=initMap-->`
Loaded the OMS after it, and used jQuery's document.ready to call initmap myself.
I suppose you could do an equivalent to document.ready if you dont want to load jquery.

**Comment by SatdayaStudios** [comment]
There are benefits to async defer for data-heavy sites that cannot be ignored.

http://peter.sh/experiments/asynchronous-and-deferred-javascript-execution-explained/

Based on the explanation above I was able to resolve this issue by simply deferring the loading of OMS until after the full DOM has loaded. Like so:

`
<script async defer src="https://maps.googleapis.com/maps/api/js?key=_key_"></script>
<script defer src="https://yoursite.com/assets/plugins/oms.min.js"></script>`

**Comment by erics-wows** [comment]
Running into the same "Uncaught ReferenceError: OverlappingMarkerSpiderfier is not defined" issue as well. Have tried the suggestions above, but no luck.

Setup is as follows: 
<head> has Async GoogleMap API load is followed by oms. GoogleMap API callback calls initMap(), which references OverlappingMarkerSpiderfier(), a follows:
<script async defer src="https://maps.googleapis.com/maps/api/js?key=OurApiKey&signed_in=true&callback=initMap&libraries=geometry"></script>
<script defer src="oms.min.js"></script>

In <body>, have a <script> which includes the 
function initMap() {
	var gm = google.maps;
        wMap.centerPt.ll = new google.maps.LatLng(wMap.centerPt.latLong)      
        map = new gm.Map(document.getElementById('map'), {
			zoom: 7,
			scaleControl: true,
			center: wMap.centerPt.ll,
			mapTypeControl: true
        });	
	// SPIDERFY CODE
	oms = new OverlappingMarkerSpiderfier(map);
        iw = new gm.InfoWindow();
	oms.addListener('click', function(marker, event) {
	    iw.setContent(marker.desc);
	    iw.open(map, marker);
        });
        //=========

Any suggestions or example code using a similar working setup would be most appreciated.
Eric

**Comment by grampy** [comment]
Devised a solution for this scenario.  I am dynamically loading Google Maps API in a geocoding application  using a Maps API key provided in the querystring of the page URL and then in the call back dynaimcally loading OverlappingMarkerSpiderfier and once it has loaded initialise my map and oms.

```javascript
<script>
        var  infoWin, map, oms, options,timer=0;

	// parse query string parameters into 'options' object
	document.location.search.replace(/([^?=&]+)(=([^&#]*))?/g, function($0, $1, $2, $3) {options[$1] = $3;});
	
	// load googlemaps API dynamically with Javascript API key provided by caller in URL query string (key=)
	var script_tag = document.createElement('script');
	script_tag.setAttribute("type","text/javascript");
	script_tag.setAttribute("src","https://maps.google.com/maps/api/js?" + (options.key ? "key=" + options.key + "&" : "") + "callback=initMap");
	(document.getElementsByTagName("head")[0] || document.documentElement).appendChild(script_tag);

	function initMap(){		// N.B. called via callback parameter when googlemaps API loaded
	
		// Load OMS dynamically now that Google Maps API has loaded
		var script_tag = document.createElement('script');
                script_tag.setAttribute("type","text/javascript");
                script_tag.setAttribute("src","shared/js/oms.min.js");
                (document.getElementsByTagName("head")[0] || document.documentElement).appendChild(script_tag);	
		// initiate loading of oms and initialise map
		setTimeout(loadMap, 50);
	};
	
	function loadMap() {
	
		// initialise Google map and spiderfier once oms (spiderfier) loaded

		if (OverlappingMarkerSpiderfier) {
			map = new google.maps.Map(canvas, {
				zoom: 2,
				center: {lat: 51, lng: 0}
			});
			infoWindow = new google.maps.InfoWindow();
			geocoder = new google.maps.Geocoder();
			
			oms = new OverlappingMarkerSpiderfier(map);

			oms.addListener('click', function(marker, event) {
				infoWin.call(marker);
			});		
		} else {
			if (timer < 1000) {
				timer += 50;
				setTimeout(loadMap, 50);
			} else {
				alert('OverlappingMarkerSpiderfier taking too long to load');
				throw('giving up!')
			}
		}
	};
	function infoWin() {
		infoWindow.setContent(	'name: '+this.title));
		infoWindow.open(map, this);
	};
</script>	
```

**Comment by jawj** [comment]
That looks workable, although the need to load the libraries sequentially is regrettable. I'm working on v1.0 of OMS (thanks to some external sponsorship), and hoping to have it ready in the next few weeks. Async loading is one of the things I'll be addressing.
